### DISCUSSION TITLE
SVG Composite Appearing Shifted With Transparent Elements

### DISCUSSION THREAD

**Discussion by jrock2468** [ImageMagick/ImageMagick#2780] [Help]
I'm fairly new to ImageMagick.  I am compositing one SVG on top of a PNG, but the resulting PNG shows a shift in the SVG's x-axis that I wasn't expecting.  The SVG is made up of a helicopter curve and a transparent rectangle because I don't want to center the helicopter in the resulting graphic.

Any ideas how I use ImageMagick to composite something that looks more like "Helicopters-11-dark (desired).png" than the actual output I get?  SVG, PNG, resulting PNG, and desired PNG, and two Affinity Designer screenshot showing the dimensions of the Helicopter curve and Rectangle are attached.

My ImageMagick script (bash) is:
```
magick composite -compose atop -gravity north -geometry +0+220 \
      -background transparent Helicopters-11.svg\
      Watchface-Background-Dark.png \
      Helicopters-11-dark.png
```

Thanks in advance.
[Composite Problem Files.zip](https://github.com/ImageMagick/ImageMagick/files/5467659/Composite.Problem.Files.zip)

**Comment by fmw42** [comment]
I cannot seem to open/view your svg file. It might not be a valid file or may be corrupt.

**Reply by jrock2468** [reply]
Are you getting an error or just not seeing an image?  The image itself is white with a transparent background.  Does this one work better for you (I made it a different stroke and fill color)?
[Helicopters-11.svg.zip](https://github.com/ImageMagick/ImageMagick/files/5467879/Helicopters-11.svg.zip)

**Comment by fmw42** [comment]
try removing the transformation section at the top. When removed the helicopter is shifted to the left rather than the center.

**Reply by jrock2468** [reply]
Thanks for the idea.  The transformation is part of a larger project where this operation occurs on hundreds of files.  The transformation is appropriate for how the files were originally designed.

**Comment by snibgo** [comment]
You seem to have added padding within the SVG to get the result you want, but it isn't quite what you want.

If you want the image shifted 10 pixels to the right, you can change the geometry to:
```
-geometry +10+220
```
But it might make more sense to remove the padding from within the SVG and do it entirely in geometry (or do it all in the SVG so there is no need for geometry).

**Reply by jrock2468** [reply]
Thanks for the idea, but this is part of a larger project where I'm utilizing batch scripting to get away from manual adjustments.  The transformation of the SVG is appropriate (in terms of pixels up/down/left/right).

**Comment by jrock2468** [comment]
I did a lot more digging, found a partial solution, and some other questions for those with more experience.

It seems that if I make a PNG intermediate file with magick, and composite between two PNGs, I get what I want.

```
magick convert -background transparent Helicopter-11.svg\
      "$OUTPUT_FOLDER/Helicopter-11-magick.png"

magick composite -compose atop -gravity north -geometry +0+220 \
      -background transparent "$OUTPUT_FOLDER/Helicopter-11-magick.png" \
      $DARKBG \
      "$OUTPUT_FOLDER/Helicopter-11-dark-from-svg-and-png.png"
```
![Helicopters-11-dark-from-svg-and-png](https://user-images.githubusercontent.com/23647639/97770087-6137e300-1afe-11eb-9716-986e8f08ccc9.png)

If I composite between the two SVGs directly, I get an undesired shift in the helicopter SVG.  It's important to note that the helicopter SVG has a transparent rectangle to create the desired spacing.

```
composite -compose atop -gravity north -geometry +0+220 \
    -background transparent Helicopter-11.svg\
    Watchface-Background-Dark.png\
    "$OUTPUT_FOLDER/Helicopter-11-dark-from-svgs.png"
```

![Helicopters-11-dark-from-transparent-rectangle-svg](https://user-images.githubusercontent.com/23647639/97770162-e7542980-1afe-11eb-92fa-e3e7f3bf1e55.png)

If I composite between the two SVGs, but I alter the Helicopter SVG to have a one pixel stroke rectangle instead of transparent, then Imagemagick behaves like I expect.  I've attached this SVG below (Zip'd).

```
composite -compose atop -gravity north -geometry +0+220 \
      -background transparent Helicopter-11-onepixel-rectangle.svg \
      Watchface-Background-Dark.png\
      "$OUTPUT_FOLDER/Helicopter-11-dark-from-onepixel-rectangle-svg.png"
```

[Helicopters-11-onepixel-rectangle.svg.zip](https://github.com/ImageMagick/ImageMagick/files/5468462/Helicopters-11-onepixel-rectangle.svg.zip)
![Helicopters-11-dark-from-onepixel-rectangle-svg](https://user-images.githubusercontent.com/23647639/97770215-5467bf00-1aff-11eb-90f4-9d9f5179e8b5.png)

I'm not sure why ImageMagick is handling the transparent rectangle SVG file differently than with the intermediate PNG, but I'll just use the two step process to keep moving forward.  Hopefully these descriptions helps someone else out there in the future, though.
